### PR TITLE
Add resource Monitor UI

### PR DESCRIPTION
**Pull request by yuxuan-ms**
## Bug / Requirement Description
Clearly and concisely describe the problem.

## Solution description
Describe your code changes in detail for reviewers.

## Checklist:
- [ ] Test
- [ ] Example (both test_plan.py and .rst)
- [ ] Documentation (API)
- [ ] News fragment present for release notes
- [ ] MS info leakage check
- [ ] For new driver: driver index page
- [ ] For new assertion: ui/pdf/std renderers, documentation
- [ ] For new cmdline arg: documentation

### CODE REVIEW

File: testplan/testing/base.py [hunk 1]
```diff
@@ -83,6 +85,9 @@ def get_options(cls):
             ConfigOption("after_start", default=None): start_stop_signature,
             ConfigOption("before_stop", default=None): start_stop_signature,
             ConfigOption("after_stop", default=None): start_stop_signature,
+            ConfigOption(
+                "result", default=result.Result
+            ): validation.is_subclass(result.Result),
```
Comment (Pyifan) [inline testplan/testing/base.py:90]: this is a bit strange, multitest result as test config option?

File: testplan/exporters/testing/json/base.py [hunk 2]
```diff
@@ -133,6 +152,14 @@ def export(
             data = test_plan_schema.dump(source)
             attachments_dir = json_path.parent / ATTACHMENTS
 
+            # Save resource monitor data
+            if source.resource_meta_path:
+                resource_dir = json_path.parent / RESOURCE_DATA
+                save_resource_data(source, resource_dir)
+                self.logger.user_info(
+                    "Resource data has been saved in %s", resource_dir
```
Comment (Pyifan) [inline testplan/exporters/testing/json/base.py:160]: Resource monitor data has been saved in ...

**Review by Pyifan — CHANGES_REQUESTED** [review]

File: testplan/monitor/resource.py [hunk 3]
```diff
@@ -106,15 +134,28 @@ def collect_cpu_usage(self) -> float:
     def collect_memory_usage(self) -> int:
         return self.memory_size - psutil.virtual_memory().available
 
+    @staticmethod
+    def _ensure_positive(num):
```
Comment (M6AI) [inline testplan/monitor/resource.py:138]: This is a bit misleading. We ensure non-negative not positive. Also I think it is cleaner to write return max(x, 0). Third thing: do you need to ensure integer as well?

File: testplan/runners/pools/remote.py [hunk 4]
```diff
@@ -74,6 +74,10 @@ def __init__(self, **options) -> None:
             options["pool_type"] = "thread"
         super().__init__(**options)
 
+    @property
+    def host(self) -> str:
```
Comment (M6AI) [inline testplan/runners/pools/remote.py:78]: Perhaps name it `remote_host`? No strong opinion here, up to your decision, just suggesting.

File: testplan/monitor/resource.py [hunk 5]
```diff
@@ -43,11 +44,34 @@ class ProcessResourceData(ResourceData):
     pid: int
 
 
+class HostResourceRow(NamedTuple):
```
Comment (M6AI) [inline testplan/monitor/resource.py:47]: Posting here, but there is a lack of docstrings for new classes and methods/functions added in the PR.

**Review by M6AI — COMMENTED** [review]
Left a couple of observations.

File: testplan/runners/pools/remote.py (hunk 4 above)
Comment (yuxuan-ms) [reply testplan/runners/pools/remote.py:78]: Local pool uses this attribute, too

**Review by Pyifan — APPROVED** [review]

**Pull request merged**